### PR TITLE
Upgrade controller to Kustomize v4

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR brings kustomize-controller on pair with Kustomize v4:
- Upgrade `sigs.k8s.io/kustomize/api` from `v0.7.4` to `v0.8.10` (fix: #308)
- Upgrade `sigs.k8s.io/kustomize/kyaml` to `v0.10.21`
- Upgrade `sigs.k8s.io/cli-utils` to `v0.25.1-0.20210608181808-f3974341173a`
- Upgrade `sigs.k8s.io/controller-runtime` to `v0.9.0` (ref: https://github.com/fluxcd/flux2/issues/1486)
- Upgrade `k8s.io/client-go` to `0.21.1`

## Breaking changes

The major breaking change in v4 is that the set of URLs accepted by kustomize in the `resources` filed is reduced to only file system paths or values compatible with git clone. This means you can no longer use resources from archives (zip, tgz, etc) nor S3, GCS, Minio, Mercurial, etc.

More breaking changes here: https://github.com/fluxcd/flux2/issues/918

## Kustomize panic workaround

This PR comes with a workaround to a kyaml bug (fix: #341, fix: #310) by serialising the kustomize build runs to avoid the OpenAPI concurrent map read/write panic kubernetes-sigs/kustomize#3659.

Note that this workaround comes with a severe performance penalty.

## Known Kustomize v4 bugs

After this upgrade the controller will be affected by these Kustomize bugs:
- If you have non-string keys in your manifests, the kustomize-controller will fail to reconcile https://github.com/kubernetes-sigs/kustomize/issues/3446
- ~If you have non-ASCII characters (e.g. Chinese comments) in your YAMLs the kustomize-controller will panic and crash loop kubernetes-sigs/kustomize#3605~ Fixed in kustomize-controller by using a newer version of go-yaml

## Test image

To test this PR please use the following image:

```
ghcr.io/fluxcd/kustomize-controller:kustomize-v4-rc-2ada9f21
```

To upgrade kustomize-controller on an cluster bootstrapped with `flux`, add the following `images` section to your `flux-system/kustomization.yaml`:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
  - gotk-components.yaml
  - gotk-sync.yaml
images:
- name: ghcr.io/fluxcd/kustomize-controller
  newName: ghcr.io/fluxcd/kustomize-controller
  newTag: kustomize-v4-rc-2ada9f21
```

### CODE REVIEW

**Comment by kingdonb** [comment]
I ran into kubernetes-sigs/kustomize#3446 and echo the sentiment in https://github.com/kubernetes-sigs/kustomize/issues/3446#issuecomment-761764629

It's basically the same use case as described in your report there, integer key for use in a HelmRelease triggers this issue -- my nginx ingress opens a TCP port 2222 to accommodate an extra service in the load balancer, ssh for `deis-builder` service, it does this by adding a section in values that looks like this:

```
  values:
    tcp:
      2222: "deis/deis-builder:2222"
```

The output looks like this:

```
80s         Normal   error               kustomization/flux-system                        kustomize build failed: map[string]interface {}{"apiVersion":"helm.toolkit.fluxcd.io/v2beta1", "kind":"HelmRelease", "metadata":map[string]interface {}{"name":"ingress-nginx", "namespace":"ingress-nginx"}, "spec":map[string]interface {}{"chart":map[string]interface {}{"spec":map[string]interface {}{"chart":"ingress-nginx", "interval":"1m", "sourceRef":map[string]interface {}{"kind":"HelmRepository", "name":"ingress-nginx", "namespace":"flux-system"}, "version":"3.23.0"}}, "interval":"5m", "values":map[string]interface {}{"controller":map[string]interface {}{"extraArgs":map[string]interface {}{"enable-ssl-passthrough":""}, "kind":"DaemonSet", "service":map[string]interface {}{"nodePorts":map[string]interface {}{"http":32082, "https":32443}, "type":"NodePort"}}, "tcp":map[interface {}]interface {}{2222:"deis/deis-builder:2222"}}}}: json: unsupported type: map[interface {}]interface {}
```

Basically no way I would have understood that error message without following the links, and I'm not sure of any way to work around it. This would block me from upgrading.

**Comment by stefanprodan** [comment]
@kingdonb try:

```yaml
values:
    tcp:
      "2222": "deis/deis-builder:2222"
```

**Comment by kingdonb** [comment]
That sorted it for me, perfect 👍 no other issues at this time

**Comment by stefanprodan** [comment]
@kingdonb Given the nginx ingress popularity, I imagine this will impact many Flux users. Indeed that error doesn't say much, but now the Kubernetes object is embedded in the error message so at least you know exactly which object caused the issue.

**Comment by nitinpatil1992** [comment]
Thank you for the PR.

Bootstrapping new cluster with this PR gives kustomize build failed error as raised per [issue -308](https://github.com/fluxcd/kustomize-controller/issues/308)

```
$ kustomize-controller git:(kustomize-v4) kg deploy -n flux-system kustomize-controller -o jsonpath="{.spec.template.spec.containers[0].image}"
ghcr.io/fluxcd/kustomize-controller:kustomize-v4-rc-de0310e7

$ kubectl logs -lapp=kustomize-controller -n flux-system -f
...
{"level":"info","ts":"2021-05-23T23:36:03.151Z","logger":"controller.kustomization","msg":"Starting workers","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","worker count":4}
{"level":"info","ts":"2021-05-23T23:36:03.161Z","logger":"controller.kustomization","msg":"Source is not ready, artifact not found","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"sandboxuk-spot","namespace":"flux-system"}
{"level":"error","ts":"2021-05-23T23:36:04.160Z","logger":"controller.kustomization","msg":"Reconciliation failed after 1.008442786s, next try in 10m0s","reconciler group":"kustomize.toolkit.fluxcd.io","reconciler kind":"Kustomization","name":"flux-system","namespace":"flux-system","revision":"main/a3b95a02fe75bb477a01355b7cbc0ea23ac77723","error":"kustomize build failed: accumulating resources: accumulation err='accumulating resources from './clusters/sandboxuk-spot': read /tmp/flux-system534296925/clusters/sandboxuk-spot: is a directory': recursed merging from path '/tmp/flux-system534296925/clusters/sandboxuk-spot': may not add resource with an already registered id: ~G_v1_Namespace|~X|flux-system"}
```

Tried removing GitRepository object as per logs but didn't help out!

**Comment by stefanprodan** [comment]
@nitinpatil1992 do you have use `$patch: delete` to remove the `flux-system` namespace? I don't see how that error is related to #308, can you provide an example repo please?

**Comment by nitinpatil1992** [comment]
@stefanprodan I didn't patch/delete flux-system namespace. 
Here is my repo structure:
```
bootstrap-repo
├── README.md
├── bases
│   ├── gotk-components.yaml
│   ├── gotk-sync.yaml
│   └── kustomization.yaml
└── clusters
    └── sandboxuk-spot
        ├── infrastructure.yaml
        └── kustomization.yaml
```
Files from `bases` are from flux v2 bootstrap document(flux namespace and all CRDs).
My custom files are as follows 
```
#clusters/sandboxuk-spot/kustomization.yaml
kind: Kustomization
bases:
- ./../../bases
resources:
- infrastructure.yaml

# clusters/sandboxuk-spot/infrastructure.yaml
---
apiVersion: source.toolkit.fluxcd.io/v1beta1
kind: GitRepository
metadata:
  name: sandboxuk-spot
  namespace: flux-system
spec:
  interval: 2m0s
  ref:
    branch: kustomize
  url: ssh://git@github.com/<org>/<infrastructure-repo>
---
apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
metadata:
  name: sandboxuk-spot
  namespace: flux-system
spec:
  interval: 5m0s
  path: ./kustomize/sandboxuk-spot.yaml
  prune: true
  sourceRef:
    kind: GitRepository
    name: sandboxuk-spot
  validation: client
```
The thing is bootstrap worked very well for the first time installation.
Then I switched the directory structure and these duplicate resources errors came into picture.

**Comment by nitinpatil1992** [comment]
`./kustomize/sandboxuk-spot.yaml` path is from infrastructure repo. Isn't that path refer to GitRepository Objects repo?

```
  sourceRef:
    kind: GitRepository
    name: sandboxuk-spot
```

**Comment by stefanprodan** [comment]
@nitinpatil1992 Flux Kustomization does not accept a file as path, but a dir, please see the docs I posted above.

**Comment by ash2k** [comment]
Re. `cli-utils` - I'm trying to bump the dependencies here https://github.com/kubernetes-sigs/cli-utils/pull/361. I wonder if this would fix the issues in this PR.

**Comment by stefanprodan** [comment]
@ash2k that you so much for this, I've switched to your PR and all tests are passing 🍻

**Comment by ash2k** [comment]
FYI @stefanprodan the PR has been merged, so this one should be unblocked now. Cheers.

**Comment by stefanprodan** [comment]
@ash2k I'm using the master branch and works great, I'll keep an eye on for the next cli-utils semver release. Thanks again for helping out with this, it wound've been a major pain for us to maintain a fork of kstatus.

**Comment by bbriggs** [comment]
@stefanprodan I had the same issue and this patch resolved it for me as well. Any ETA on when it might land in prod?

**Comment by stefanprodan** [comment]
@bbriggs what issue are you referring to?

> Any ETA on when it might land in prod?

I'm waiting for controller-runtime v0.9 release. As documented in the PR I had to pin controller-runtime to `v0.9.0-beta` due to breaking changes in Kubernetes client-go.

**Comment by bbriggs** [comment]
> @bbriggs what issue are you referring to?

the Kustomize panic when applying things with `$patch: delete`

**Comment by bbriggs** [comment]
> I'm waiting for controller-runtime v0.9 release. As documented in the PR I had to pin controller-runtime to `v0.9.0-beta` due to breaking changes in Kubernetes client-go.

Thank you for your hard work on this! It was about to be a showstopper for us and publishing the workaround image in a PR allowed us to verify the bug and move past it so we could work on other problems. Not many folks do that, so we're very grateful :+1: :tada:

**Review by hiddeco — APPROVED** [review]
LGTM, thank you @stefanprodan :100:
